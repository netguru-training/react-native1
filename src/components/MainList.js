import React from 'react';
import {
  StyleSheet,
  Text,
  View,
  ListView,
  TextInput,
  TouchableOpacity,
  AsyncStorage,
  LayoutAnimation
} from 'react-native';
import { StackNavigator } from 'react-navigation';
import { Item, Input } from 'native-base';
import dummyData from './dummyData';
import TaskList from './TaskList';

export default class MainList extends React.Component {
  state = {
    toDoItems: [],
    doneItems: [],
    item: {}
  };

  static navigationOptions = {
    title: 'Tasks'
  };

  componentWillMount() {
    this.setState({ toDoItems: dummyData });
  }

  onChangeText = text => {
    this.setState({
      item: { name: text, id: Math.random(), desc: '', isCompleted: false }
    });
  };

  onNewItem = e => {
    const arr = [this.state.item, ...this.state.toDoItems];
    LayoutAnimation.spring();
    this.setState({
      toDoItems: arr,
      item: {}
    });
  };

  checkItemWithId = id => {
    const allItems = [...this.state.toDoItems, ...this.state.doneItems];
    const itemsAfterChecking = allItems.map(item => {
      if (item.id === id) {
        item.isCompleted = !item.isCompleted;
      }
      return item;
    });

    const newToDoList = itemsAfterChecking.filter(item => !item.isCompleted);
    const newDoneList = itemsAfterChecking.filter(item => item.isCompleted);

    LayoutAnimation.spring();
    this.setState({ toDoItems: newToDoList, doneItems: newDoneList });
  };

  editDescription = (description, id) => {
    const allItems = [...this.state.toDoItems, ...this.state.doneItems];
    const item = allItems.map(item => {
      item.id === id
        ? (item = {
            id: item.id,
            name: item.name,
            description: description,
            isCompleted: item.isCompleted
          })
        : item;
    });

    const arr = [item, ...this.state.items];
    this.setState({
      items: arr
    });
  };

  moveToScreen = id => {
    const allItems = [...this.state.toDoItems, ...this.state.doneItems];
    const item = allItems.filter(item => item.id === id)[0];
    console.log('wybrany item', item);
    this.props.navigation.navigate('TaskFull', {
      item: item,
      editDescription: this.editDescription
    });
  };

  removeItemWithId = id => {
    var allItems = [...this.state.toDoItems, ...this.state.doneItems];
    const item = allItems.filter(item => item.id === id)[0];

    const index = allItems.indexOf(item);
    allItems.splice(index, 1);
    console.log('allItems', allItems);
    console.log('item', item);
    console.log('index', index);
    console.log('allitems', allItems)

    const newToDoList = allItems.filter(item => !item.isCompleted);
    const newDoneList = allItems.filter(item => item.isCompleted);

    LayoutAnimation.spring();
    this.setState({ toDoItems: newToDoList, doneItems: newDoneList });
  };

  render() {
    return (
      <View style={{ flex: 1 }}>
        <View style={{ paddingLeft: 4, paddingTop: 4, paddingRight: 4 }}>
          <Item regular>
            <Input
              style={{ backgroundColor: '#fff' }}
              onSubmitEditing={this.onNewItem}
              placeholder="Enter Your New Task"
              returnKeyType="done"
              onChangeText={this.onChangeText}
              value={this.state.item.name}
            />
          </Item>
        </View>
        <TaskList
          style={{ flex: 1 }}
          toDoItems={this.state.toDoItems}
          doneItems={this.state.doneItems}
          checkItemWithId={this.checkItemWithId}
          editDescription={this.editDescription}
          moveToScreen={this.moveToScreen}
          removeItemWithId={this.removeItemWithId}
        />
      </View>
    );
  }
}
