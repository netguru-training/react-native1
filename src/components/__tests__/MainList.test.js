import { LayoutAnimation } from "react-native";
import React from "react";
import MainList from "../MainList";

import { shallow } from "enzyme";
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

const wrapper = shallow(<MainList />);
const instance = wrapper.instance();
const spyOnChangeText = jest.spyOn(instance, "onChangeText");
const spyOnCheckItemWithId = jest.spyOn(instance, "checkItemWithId");
const spyOnNewItem = jest.spyOn(instance, "onNewItem");
const spyOnEditDescription = jest.spyOn(instance, "editDescription");
const spyOnEditName = jest.spyOn(instance, "editName");
const spyOnMoveToScreen = jest.spyOn(instance, "moveToScreen");
const spyOnRemoveItemWithId = jest.spyOn(instance, "removeItemWithId");
const layoutAnimation = jest.mock("LayoutAnimation", () => ({
  spring: jest.fn()
}));

describe("Rendering", () => {
  test("Component with default props", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
describe("Actions", () => {
  test("When I pass XXXX to onChangeText function, I should get item state XXXX", () => {
    instance.onChangeText("XXXX");
    expect(spyOnChangeText).toHaveBeenCalled();
    expect(wrapper.state().item.name).toBe("XXXX");
  });
  test("When I call the onNewItem function, it adds the item to toDoList", () => {
    instance.onNewItem();
    expect(wrapper.state().toDoItems.length).toBe(1);
  });
  test("When I pass 0 to checkItemWithId function, the item with passed it should be moved to doneItems list.", () => {
    instance.checkItemWithId(wrapper.state().toDoItems[0].id);
    expect(wrapper.state().doneItems.length).toBe(1);
  });
  test("When I call editDescription function, it finds the correct item and edit description in it", () => {
    instance.onChangeText("XXXX");
    instance.onNewItem();
    const id = wrapper.state().toDoItems[0].id;
    const desc = "wall";
    const element = wrapper.state().toDoItems.find(item => item.id === id);
    instance.editDescription(desc, id);
    expect(spyOnEditDescription).toHaveBeenCalled();
    expect(element.desc).toBe(desc);
  });
  test("When I call editName function, it finds the correct item and edit name in it", () => {
    const id = wrapper.state().toDoItems[0].id;
    const name = "manamana";
    const element = wrapper.state().toDoItems.find(item => item.id === id);
    instance.editName(name, id);
    expect(spyOnEditName).toHaveBeenCalled();
    expect(element.name).toBe(name);
  });
  test("When I call moveToScreen function, I should open taskFull screen with current Id", () => {
    const id = wrapper.state().toDoItems[0].id;
    const wrapperNew = shallow(
      <MainList navigation={{ navigate: jest.fn() }} />
    );
    const instanceNew = wrapperNew.instance();
    instanceNew.moveToScreen(id);
  });
  test("When I call removeItemWithId function, it should remove the task", () => {
    const id =wrapper.state().toDoItems[0].id;
    expect(wrapper.state().toDoItems.length).toBe(1);
    instance.removeItemWithId(id);
    expect(wrapper.state().toDoItems.length).toBe(0);
  });
});
