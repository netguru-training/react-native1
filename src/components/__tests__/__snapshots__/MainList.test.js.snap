// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering Component with default props 1`] = `
<Spinner
  size="large"
/>
`;
